### ISSUE TITLE
New Explanation Multiplicity Metrics for EBM

### ISSUE THREAD

**Issue by athyrson06** [CaioRhoden/tfm-distill-multiplicity#2]
import numpy as np
import pandas as pd

from itertools import combinations
from scipy.stats import spearmanr

from interpret.glassbox import ExplainableBoostingClassifier
from sklearn.metrics import brier_score_loss
from sklearn.model_selection import KFold

from recgam.utils import calculate_disagreement, calculate_ambiguity

def calculate_global_explanation_multiplicity(importances_matrix):
    """
    Measures disagreement between models in the ranking of
    global term importances using pairwise Spearman correlation.
    """

    n_models = importances_matrix.shape[0]

    spearman_corrs = []

    for m1, m2 in combinations(range(n_models), 2):
        corr, _ = spearmanr(
            importances_matrix[m1],
            importances_matrix[m2]
        )

        if not np.isnan(corr):
            spearman_corrs.append(corr)

    return {
        "mean_spearman_correlation": (
            np.mean(spearman_corrs)
            if spearman_corrs
            else np.nan
        )
    }

# ============================================================
# 2. FUNCTIONAL MULTIPLICITY
# ============================================================

def calculate_functional_multiplicity(term_scores_list, term_features_list=None):
    n_models = len(term_scores_list)
    fed_by_order = {1: [], 2: [], 3: []}
    fed_scores = []

    for m1, m2 in combinations(range(n_models), 2):
        model_1_terms = term_scores_list[m1]
        model_2_terms = term_scores_list[m2]
        n_terms = min(len(model_1_terms), len(model_2_terms))
        pair_fed = []

        for term_idx in range(n_terms):
            s1 = np.asarray(model_1_terms[term_idx])
            s2 = np.asarray(model_2_terms[term_idx])
            if s1.shape != s2.shape:
                continue

            term_range = max(np.ptp(s1), np.ptp(s2))
            if term_range < 1e-12:
                continue

            normalized_diff = np.mean(np.abs(s1 - s2)) / term_range
            pair_fed.append(normalized_diff)

            if term_features_list is not None:
                order = len(term_features_list[m1][term_idx])
                fed_by_order.setdefault(order, []).append(normalized_diff)

        if pair_fed:
            fed_scores.append(np.mean(pair_fed))

    result = {
        "mean_normalized_fed": np.mean(fed_scores) if fed_scores else np.nan
    }
    for order, scores in fed_by_order.items():
        if scores:
            result[f"mean_fed_order{order}"] = np.mean(scores)
    return result

# ============================================================
# 3. LOCAL EXPLANATION MULTIPLICITY
# ============================================================

def calculate_local_explanation_multiplicity(local_contribs_matrix):
    """
    Mean pairwise Local Attribution Discrepancy (LAD).
    """

    n_models = local_contribs_matrix.shape[0]

    lad_scores = []

    for m1, m2 in combinations(range(n_models), 2):

        lad = np.mean(
            np.abs(
                local_contribs_matrix[m1]
                - local_contribs_matrix[m2]
            )
        )

        lad_scores.append(lad)

    return {
        "mean_local_attribution_discrepancy": (
            np.mean(lad_scores)
            if lad_scores
            else np.nan
        )
    }

# ============================================================
# 4. EXPLANATION AGREEMENT (NOVO)
# ============================================================

def calculate_explanation_agreement(
    clf_list,
    X_data,
    k=3
):
    """
    Calculates Jaccard similarity between top-k features
    identified by different models for each instance.

    Parameters:
    -----------
    clf_list : list
        List of trained EBM classifiers
    X_data : array-like
        Data to explain
    k : int
        Number of top features to consider

    Returns:
    --------
    dict with mean Jaccard similarity across all instances and model pairs
    """

    n_models = len(clf_list)
    n_instances = X_data.shape[0]

    # Get local explanations for each model
    local_explanations = []

    for clf in clf_list:
        local_expl = clf.explain_local(X_data)
        local_explanations.append(local_expl)

    # Calculate pairwise Jaccard similarity
    jaccard_scores = []

    for m1, m2 in combinations(range(n_models), 2):
        for instance_idx in range(n_instances):

            # Get top-k features for model 1
            expl1 = local_explanations[m1]
            data1 = expl1.data(instance_idx)
            scores1 = np.abs(data1["scores"])
            top_k_1 = set(
                np.argsort(scores1)[-k:]
            )

            # Get top-k features for model 2
            expl2 = local_explanations[m2]
            data2 = expl2.data(instance_idx)
            scores2 = np.abs(data2["scores"])
            top_k_2 = set(
                np.argsort(scores2)[-k:]
            )

            # Calculate Jaccard similarity
            intersection = len(top_k_1 & top_k_2)
            union = len(top_k_1 | top_k_2)

            if union > 0:
                jaccard = intersection / union
                jaccard_scores.append(jaccard)

    return {
        f"mean_jaccard_top{k}": (
            np.mean(jaccard_scores)
            if jaccard_scores
            else np.nan
        )
    }

# ============================================================
# 5. EXPLANATION SPARSITY / COMPLEXITY (NOVO)
# ============================================================
def _nd_roughness(term_scores):
    """Mean abs second difference, averaged across all axes present."""
    term_scores = np.asarray(term_scores)
    roughnesses = []
    for axis in range(term_scores.ndim):
        if term_scores.shape[axis] > 2:
            second_diff = np.diff(term_scores, n=2, axis=axis)
            roughnesses.append(np.mean(np.abs(second_diff)))
    return np.mean(roughnesses) if roughnesses else np.nan

def calculate_explanation_sparsity(clf):
    """
    Measures sparsity and complexity of EBM explanations,
    broken down by interaction order (1 = main effect,
    2 = pairwise, 3 = triple interaction).
    """

    global_expl = clf.explain_global()
    global_data = global_expl.data()

    importances = np.abs(global_data["scores"])
    n_active_features = np.sum(importances > 1e-6)
    n_total_features = len(importances)
    sparsity_ratio = 1 - (n_active_features / n_total_features)

    n_pairwise = 0
    n_triple = 0
    if hasattr(clf, 'term_features_'):
        for term_features in clf.term_features_:
            if len(term_features) == 2:
                n_pairwise += 1
            elif len(term_features) == 3:
                n_triple += 1

    roughness_by_order = {1: [], 2: [], 3: []}
    if hasattr(clf, 'term_scores_') and hasattr(clf, 'term_features_'):
        for term_scores, term_features in zip(clf.term_scores_, clf.term_features_):
            r = _nd_roughness(term_scores)
            if not np.isnan(r):
                order = len(term_features)
                roughness_by_order.setdefault(order, []).append(r)

    all_roughness = [r for scores in roughness_by_order.values() for r in scores]

    result = {
        "n_active_features": n_active_features,
        "n_total_features": n_total_features,
        "sparsity_ratio": sparsity_ratio,
        "n_interaction_terms": n_pairwise + n_triple,   # backward-compat combined count
        "n_pairwise_interaction_terms": n_pairwise,
        "n_triple_interaction_terms": n_triple,
        "mean_roughness": np.mean(all_roughness) if all_roughness else 0.0,
        "n_terms": len(clf.term_scores_) if hasattr(clf, 'term_scores_') else 0,
    }
    for order, scores in roughness_by_order.items():
        if scores:
            result[f"mean_roughness_order{order}"] = np.mean(scores)
    return result

**Comment by athyrson06** [comment]
usage example # ============================================================
# 6. EXPERIMENT CONFIGURATION
# ============================================================

n_seeds = 5

kf = KFold(
    n_splits=2,
    shuffle=True,
    random_state=34
)

metrics_results = []


# ============================================================
# 7. CROSS-VALIDATION
# ============================================================

for fold, (train_index, val_index) in enumerate(
    kf.split(X_train)):

    print(f"\n{'=' * 30}")
    print(f"Fold {fold + 1}")
    print(f"{'=' * 30}")

    # --------------------------------------------------------
    # Training split
    # --------------------------------------------------------

    if isinstance(X_train, np.ndarray):
        X_ = X_train[train_index]
        X_val_fold = X_train[val_index]
    else:
        X_ = X_train.iloc[train_index]
        X_val_fold = X_train.iloc[val_index]

    if isinstance(y_train, np.ndarray):
        y_ = y_train[train_index].copy()
        y_val_fold = y_train[val_index]
    else:
        y_ = y_train.iloc[train_index].copy()
        y_val_fold = y_train.iloc[val_index]


    # --------------------------------------------------------
    # Storage
    # --------------------------------------------------------

    preds_proba_list = []
    preds_class_list = []

    importances_list = []
    term_scores_list = []
    term_features_list = []   # NEW: track term structure (order) per model
    clf_list = []

    brier_scores = []
    sparsity_metrics_list = []


    # ========================================================
    # 8. TRAIN MULTIPLE EBMs
    # ========================================================

    for seed in range(n_seeds):

        clf = ExplainableBoostingClassifier(
            interactions="2x",     # CHANGED: allow up to 3-way interactions
            max_rounds=100,
            random_state=seed
        )

        clf.fit(X_, y_)
        clf_list.append(clf)


        # ----------------------------------------------------
        # Predictions
        # ----------------------------------------------------

        y_prob = clf.predict_proba(X_val_fold)[:, 1]

        y_pred = clf.predict(X_val_fold)

        preds_proba_list.append(y_prob)
        preds_class_list.append(y_pred)

        brier_scores.append(
            brier_score_loss(
                y_val_fold,
                y_prob
            )
        )


        # ----------------------------------------------------
        # Global explanation
        # ----------------------------------------------------

        global_expl = clf.explain_global()

        global_data = global_expl.data()

        importances = np.asarray(
            global_data["scores"]
        )

        importances_list.append(importances)


        # ----------------------------------------------------
        # Functional explanation
        # ----------------------------------------------------

        term_scores_list.append(
            [
                np.asarray(term_score)
                for term_score in clf.term_scores_
            ]
        )

        term_features_list.append(   # NEW: needed for order-aware FED
            list(clf.term_features_)
        )


        # ----------------------------------------------------
        # Sparsity metrics
        # ----------------------------------------------------

        sparsity_metrics = calculate_explanation_sparsity(clf)
        sparsity_metrics_list.append(sparsity_metrics)


    # ========================================================
    # 9. CONVERT PREDICTIONS TO MATRICES
    # ========================================================

    preds_proba = np.column_stack(
        preds_proba_list
    )

    preds_class = np.column_stack(
        preds_class_list
    )


    # ========================================================
    # 10. GLOBAL IMPORTANCE MATRIX
    # ========================================================

    importances_matrix = np.vstack(
        importances_list
    )


    # ========================================================
    # 11. MULTIPLICITY METRICS
    # ========================================================

    fold_disagreement = calculate_disagreement(
        preds_class
    )[0]

    fold_ambiguity = calculate_ambiguity(
        preds_proba
    )[0].mean()

    fold_spearman = (
        calculate_global_explanation_multiplicity(
            importances_matrix
        )["mean_spearman_correlation"]
    )

    fed_result = calculate_functional_multiplicity(   # CHANGED: pass term_features_list
        term_scores_list,
        term_features_list
    )
    fold_fed = fed_result["mean_normalized_fed"]
    fold_fed_order1 = fed_result.get("mean_fed_order1", np.nan)
    fold_fed_order2 = fed_result.get("mean_fed_order2", np.nan)
    fold_fed_order3 = fed_result.get("mean_fed_order3", np.nan)


    # --------------------------------------------------------
    # Explanation Agreement
    # --------------------------------------------------------

    fold_agreement_top3 = calculate_explanation_agreement(
        clf_list,
        X_val_fold,
        k=3
    )["mean_jaccard_top3"]

    fold_agreement_top5 = calculate_explanation_agreement(
        clf_list,
        X_val_fold,
        k=5
    )["mean_jaccard_top5"]


    # --------------------------------------------------------
    # Sparsity metrics (aggregated across models)
    # --------------------------------------------------------

    mean_sparsity = np.mean([
        m["sparsity_ratio"]
        for m in sparsity_metrics_list
    ])

    mean_active_features = np.mean([
        m["n_active_features"]
        for m in sparsity_metrics_list
    ])

    mean_roughness = np.mean([
        m["mean_roughness"]
        for m in sparsity_metrics_list
    ])

    mean_n_interactions = np.mean([
        m["n_interaction_terms"]
        for m in sparsity_metrics_list
    ])

    mean_n_pairwise = np.mean([          # NEW
        m["n_pairwise_interaction_terms"]
        for m in sparsity_metrics_list
    ])

    mean_n_triple = np.mean([            # NEW
        m["n_triple_interaction_terms"]
        for m in sparsity_metrics_list
    ])

    mean_roughness_order1 = np.mean([    # NEW
        m.get("mean_roughness_order1", np.nan)
        for m in sparsity_metrics_list
    ])

    mean_roughness_order2 = np.mean([    # NEW
        m.get("mean_roughness_order2", np.nan)
        for m in sparsity_metrics_list
    ])

    mean_roughness_order3 = np.mean([    # NEW
        m.get("mean_roughness_order3", np.nan)
        for m in sparsity_metrics_list
    ])


    # ========================================================
    # 12. RESULTS
    # ========================================================

    print(
        f"Brier Score Loss (Média)      : "
        f"{np.mean(brier_scores):.4f}"
    )

    print(
        f"Prediction Disagreement       : "
        f"{fold_disagreement:.4%}"
    )

    print(
        f"Prediction Ambiguity (Var)    : "
        f"{fold_ambiguity:.4f}"
    )

    print(
        f"Expl. Rank Correlation        : "
        f"{fold_spearman:.4f}"
    )

    print(
        f"Functional Discrepancy (FED)  : "
        f"{fold_fed:.4f}"
    )

    print(
        f"  - FED (order 1, main eff.)  : "
        f"{fold_fed_order1:.4f}"
    )

    print(
        f"  - FED (order 2, pairwise)   : "
        f"{fold_fed_order2:.4f}"
    )

    print(
        f"  - FED (order 3, triple)     : "
        f"{fold_fed_order3:.4f}"
    )

    print(
        f"Explanation Agreement (Top-3) : "
        f"{fold_agreement_top3:.4f}"
    )

    print(
        f"Explanation Agreement (Top-5) : "
        f"{fold_agreement_top5:.4f}"
    )

    print(
        f"Explanation Sparsity Ratio    : "
        f"{mean_sparsity:.4f}"
    )

    print(
        f"Active Features (Média)       : "
        f"{mean_active_features:.2f}"
    )

    print(
        f"Term Roughness (Média)        : "
        f"{mean_roughness:.4f}"
    )

    print(
        f"  - Roughness (order 1)       : "
        f"{mean_roughness_order1:.4f}"
    )

    print(
        f"  - Roughness (order 2)       : "
        f"{mean_roughness_order2:.4f}"
    )

    print(
        f"  - Roughness (order 3)       : "
        f"{mean_roughness_order3:.4f}"
    )

    print(
        f"Interaction Terms (Média)     : "
        f"{mean_n_interactions:.2f} "
        f"(pairwise: {mean_n_pairwise:.2f}, "
        f"triple: {mean_n_triple:.2f})"
    )


    # ========================================================
    # 13. SAVE RESULTS
    # ========================================================

    metrics_results.append({
        "fold": fold + 1,
        "brier_score": np.mean(brier_scores),
        "disagreement": fold_disagreement,
        "ambiguity": fold_ambiguity,
        "explanation_corr": fold_spearman,
        "functional_discrepancy": fold_fed,
        "functional_discrepancy_order1": fold_fed_order1,
        "functional_discrepancy_order2": fold_fed_order2,
        "functional_discrepancy_order3": fold_fed_order3,
        "agreement_top3": fold_agreement_top3,
        "agreement_top5": fold_agreement_top5,
        "sparsity_ratio": mean_sparsity,
        "n_active_features": mean_active_features,
        "mean_roughness": mean_roughness,
        "mean_roughness_order1": mean_roughness_order1,
        "mean_roughness_order2": mean_roughness_order2,
        "mean_roughness_order3": mean_roughness_order3,
        "n_interaction_terms": mean_n_interactions,
        "n_pairwise_interaction_terms": mean_n_pairwise,
        "n_triple_interaction_terms": mean_n_triple,
    })


# ============================================================
# 14. RESULTS DATAFRAME
# ============================================================

metrics_results_df = pd.DataFrame(
    metrics_results
)

print("\nFinal results:")
metrics_results_df
